lib/kafkalib/connection.go

```diff
@@ -0,0 +1,119 @@
+package kafkalib
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"time"
+
+	awsCfg "github.com/aws/aws-sdk-go-v2/config"
+	"github.com/segmentio/kafka-go"
+	"github.com/segmentio/kafka-go/sasl/aws_msk_iam_v2"
+	"github.com/segmentio/kafka-go/sasl/scram"
+)
+
+type Mechanism string
+
+const (
+	Plain       Mechanism = "PLAIN"
+	ScramSha512 Mechanism = "SCRAM-SHA-512"
+	AwsMskIam   Mechanism = "AWS-MSK-IAM"
+)
+
+type Connection struct {
+	enableAWSMSKIAM bool
+	disableTLS      bool
+	username        string
+	password        string
+}
+
+func NewConnection(enableAWSMSKIAM bool, disableTLS bool, username, password string) Connection {
+	return Connection{
+		enableAWSMSKIAM: enableAWSMSKIAM,
+		disableTLS:      disableTLS,
+		username:        username,
+		password:        password,
+	}
+}
+
+func (c Connection) Mechanism() Mechanism {
+	if c.username != "" && c.password != "" {
+		return ScramSha512
+	}
+
+	if c.enableAWSMSKIAM {
+		return AwsMskIam
+	}
+
+	return Plain
+}
+
+func (c Connection) Dialer(ctx context.Context) (*kafka.Dialer, error) {
+	dialer := &kafka.Dialer{
+		Timeout:   10 * time.Second,
+		DualStack: true,
+	}
+
+	switch c.Mechanism() {
+	case ScramSha512:
+		mechanism, err := scram.Mechanism(scram.SHA512, c.username, c.password)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create SCRAM mechanism: %w", err)
+		}
+
+		dialer.SASLMechanism = mechanism
+		if !c.disableTLS {
+			dialer.TLS = &tls.Config{}
+		}
+	case AwsMskIam:
+		_awsCfg, err := awsCfg.LoadDefaultConfig(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load aws configuration: %w", err)
+		}
+
+		dialer.SASLMechanism = aws_msk_iam_v2.NewMechanism(_awsCfg)
+		// We don't need to disable TLS for AWS IAM since MSK will always enable TLS.
+		dialer.TLS = &tls.Config{}
+	case Plain:
+		// No mechanism
+	default:
+		return nil, fmt.Errorf("unsupported kafka mechanism: %s", c.Mechanism())
+	}
+
+	return dialer, nil
+}
+
+func (c Connection) Transport() (*kafka.Transport, error) {
```

Comment:
Is this used anywhere yet?

Reply:
No, I want to use it here though: https://github.com/artie-labs/reader/pull/434